smsotp/smsotp-authenticator/org.wso2.carbon.identity.authenticator/src/main/java/org/wso2/carbon/identity/authenticator/SMSOTPAuthenticator.java

```diff
@@ -0,0 +1,298 @@
+        newAuthenticatorProperties.put("from", prop.getProperty("from"));
+        newAuthenticatorProperties.put("text", prop.getProperty("text"));
+        context.setAuthenticatorProperties(newAuthenticatorProperties);
+
+        OneTimePassword token = new OneTimePassword();
+        String secret = OneTimePassword.getRandomNumber(SMSOTPConstants.SECRET_KEY_LENGTH);
+        otpToken = token.generateToken(secret, "" + SMSOTPConstants.NUMBER_BASE, SMSOTPConstants.NUMBER_DIGIT);
+        Object myToken = otpToken;
+        authContext.setProperty(otpToken, myToken);
+
+        Map<String, String> authenticatorProperties = context
+                .getAuthenticatorProperties();
+        clientId = authenticatorProperties
+                .get(SMSOTPConstants.API_KEY);
+        clientSecret = authenticatorProperties
+                .get(SMSOTPConstants.API_SECRET);
+        smsUrl = authenticatorProperties.get(SMSOTPConstants.SMS_URL);
+
+        String loginPage=ConfigurationFacade.getInstance().getAuthenticationEndpointURL()
+                .replace("authenticationendpoint/login.do", SMSOTPConstants.LOGIN_PAGE);
+        String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(
+                context.getQueryParams(), context.getCallerSessionKey(),
+                context.getContextIdentifier());
+        String retryParam = "";
+
+        if (context.isRetrying()) {
+            retryParam = SMSOTPConstants.RETRY_PARAMS;
+        }
+
+        try {
+            response.sendRedirect(response.encodeRedirectURL(loginPage + ("?" + queryParams)) + "&authenticators=" +
+                    getName() + retryParam);
+        } catch (IOException e) {
+            log.error("Authentication failed!", e);
+            throw new AuthenticationFailedException(e.getMessage(), e);
+        }
+
+        String username = null;
+        for (Integer stepMap : context.getSequenceConfig().getStepMap().keySet())
+            if (context.getSequenceConfig().getStepMap().get(stepMap).getAuthenticatedUser() != null &&
+                    context.getSequenceConfig().getStepMap().get(stepMap).getAuthenticatedAutenticator()
+                            .getApplicationAuthenticator() instanceof LocalApplicationAuthenticator) {
+
+                username = String.valueOf(context.getSequenceConfig().getStepMap().get(stepMap).getAuthenticatedUser());
+                break;
+            }
+        if (username != null) {
+            UserRealm userRealm = null;
+            try {
+                String tenantDomain = MultitenantUtils.getTenantDomain(username);
+                int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+                RealmService realmService = IdentityTenantUtil.getRealmService();
+                userRealm = (UserRealm) realmService.getTenantUserRealm(tenantId);
+            } catch (Exception e) {
+                throw new AuthenticationFailedException("Cannot find the user realm", e);
```

Comment:
It would be better to mention the user realm which was not found in the message itself.